File: app/src/main/java/com/originb/inkwisenote/modules/notesearch/NoteSearchActivity.java
package com.originb.inkwisenote.modules.notesearch;

import android.os.Bundle;
import android.view.View;
import android.widget.*;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import com.originb.inkwisenote.R;
import com.originb.inkwisenote.modules.backgroundjobs.BackgroundOps;
import com.originb.inkwisenote.modules.ocr.data.NoteOcrTextDao;
import com.originb.inkwisenote.modules.ocr.data.NoteOcrText;
import com.originb.inkwisenote.modules.repositories.Repositories;
import com.originb.inkwisenote.modules.repositories.SmartNotebook;
import com.originb.inkwisenote.modules.repositories.SmartNotebookRepository;
import com.originb.inkwisenote.modules.smartnotes.ui.SmartNoteGridAdapter;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class NoteSearchActivity extends AppCompatActivity {
    private EditText searchInput;
    private Button searchButton;
    private List<SmartNotebook> resultsList;

    private SmartNoteGridAdapter smartNoteGridAdapter;

    private NoteOcrTextDao noteOcrTextDao;
    private SmartNotebookRepository smartNotebookRepository;

    private RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_results);

        noteOcrTextDao = Repositories.getInstance().getNotesDb().noteOcrTextDao();
        smartNotebookRepository = Repositories.getInstance().getSmartNotebookRepository();

        searchInput = findViewById(R.id.searchInput);
        searchButton = findViewById(R.id.searchButton);

        createGridLayoutToShowNotes();

        searchButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                performSearch(searchInput.getText().toString());
            }
        });
    }

    public void createGridLayoutToShowNotes() {
        resultsList = new ArrayList<>();

        recyclerView = findViewById(R.id.note_search_card_grid_view);
        smartNoteGridAdapter = new SmartNoteGridAdapter(this, new ArrayList<>());

        recyclerView.setAdapter(smartNoteGridAdapter);
        recyclerView.setHasFixedSize(true);
    }

    private void performSearch(String query) {
        // Filter results based on query
        if (query.length() < 3) {
            Toast.makeText(this, "enter at least 3 characters to search", Toast.LENGTH_SHORT).show();
            return;
        }

        resultsList.clear();
        BackgroundOps.execute(() -> {
                    Set<SmartNotebook> smartNotebooks = smartNotebookRepository.getSmartNotebooks(query);
                    List<NoteOcrText> noteOcrs = noteOcrTextDao.searchTextFromDb(query);
                    if(noteOcrs != null && !noteOcrs.isEmpty()) {
                        Set<Long> noteIds = noteOcrs.stream()
                                .map(NoteOcrText::getNoteId)
                                .collect(Collectors.toSet());
                        smartNotebooks.addAll(smartNotebookRepository.getSmartNotebooksForNoteIds(noteIds));

                    }
                    return smartNotebooks;
                },
                smartNotebooks -> {
                    resultsList.addAll(smartNotebooks);
                    smartNoteGridAdapter.setSmartNotebooks(resultsList);
                });
    }
}

File: androidApp/src/main/java/com/originb/inkwisenote2/modules/notesearch/NoteSearchActivity.kt
package com.originb.inkwisenote2.modules.notesearch

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import com.originb.inkwisenote2.R
import com.originb.inkwisenote2.modules.repositories.SmartNotebook
import com.originb.inkwisenote2.modules.smartnotes.ui.SmartNoteGridAdapter
import org.koin.android.compat.ViewModelCompat.getViewModel

class NoteSearchActivity : AppCompatActivity() {
    private var searchInput: EditText? = null
    private var searchButton: Button? = null
    private var smartNoteGridAdapter: SmartNoteGridAdapter? = null
    private var viewModel: NoteSearchViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_results)

        // 1. Initialize ViewModel with Koin DI
        viewModel = getViewModel<NoteSearchViewModel>(this, NoteSearchViewModel::class.java)

        // 2. Initialize UI Components
        searchInput = findViewById<EditText>(R.id.searchInput)
        searchButton = findViewById<Button>(R.id.searchButton)
        setupRecyclerView()

        // 3. Set up Observers (The "Reactive" part)
        observeViewModel()

        // 4. Input events
        searchButton!!.setOnClickListener(View.OnClickListener { view: View? ->
            viewModel!!.performSearch(
                searchInput!!.getText().toString()
            )
        }
        )
    }

    private fun setupRecyclerView() {
        val recyclerView = findViewById<RecyclerView>(R.id.note_search_card_grid_view)
        smartNoteGridAdapter = SmartNoteGridAdapter(this, ArrayList<SmartNotebook>(), false)
        recyclerView.setAdapter(smartNoteGridAdapter)
        recyclerView.setHasFixedSize(true)
    }

    private fun observeViewModel() {
        // Update list when search results change
        viewModel!!.searchResults.observe(this) { results: MutableList<SmartNotebook> ->
            smartNoteGridAdapter!!.setSmartNotebooks(results.toMutableList())
        }

        // Show toast when the ViewModel sends a message
        viewModel!!.toastMessage.observe(this, Observer { message: String? ->
            if (message != null) {
                Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
            }
        })
    }
}
